File: backend/app/schemas/finance.py
import uuid
from datetime import date as date_type
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, computed_field


class GoalCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    target_amount: Decimal = Field(gt=0)
    currency: str = Field(default="RUB", min_length=3, max_length=3)


class GoalUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    target_amount: Decimal | None = Field(default=None, gt=0)


class GoalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    target_amount: Decimal
    currency: str
    achieved_at: datetime | None
    current_amount: Decimal

    @computed_field
    @property
    def progress(self) -> float:
        if self.target_amount <= 0:
            return 0.0
        return min(float(self.current_amount / self.target_amount), 1.0)


class IncomeEntryCreate(BaseModel):
    amount: Decimal = Field(gt=0)
    entry_date: date_type | None = None
    note: str | None = Field(default=None, max_length=280)
    goal_id: uuid.UUID | None = None


class IncomeEntryUpdate(BaseModel):
    goal_id: uuid.UUID | None = None
    note: str | None = Field(default=None, max_length=280)


class IncomeEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    amount: Decimal
    currency: str
    source: str
    entry_date: date_type
    note: str | None
    daily_task_instance_id: uuid.UUID | None
    goal_id: uuid.UUID | None
    created_at: datetime


class FinanceSummary(BaseModel):
    total_all_time: Decimal
    total_this_month: Decimal
    unallocated_amount: Decimal

File: mobile/lib/models/finance.dart
class FinancialGoal {
  final String id;
  final String title;
  final double targetAmount;
  final String currency;
  final DateTime? achievedAt;
  final double currentAmount;
  final double progress;

  const FinancialGoal({
    required this.id,
    required this.title,
    required this.targetAmount,
    required this.currency,
    required this.achievedAt,
    required this.currentAmount,
    required this.progress,
  });

  bool get isAchieved => achievedAt != null;

  factory FinancialGoal.fromJson(Map<String, dynamic> json) => FinancialGoal(
        id: json['id'] as String,
        title: json['title'] as String,
        targetAmount: double.parse(json['target_amount'].toString()),
        currency: json['currency'] as String,
        achievedAt: json['achieved_at'] == null ? null : DateTime.parse(json['achieved_at'] as String),
        currentAmount: double.parse(json['current_amount'].toString()),
        progress: (json['progress'] as num).toDouble(),
      );
}

class IncomeEntry {
  final String id;
  final double amount;
  final String currency;
  final String source; // "daily_task" | "manual"
  final DateTime entryDate;
  final String? note;
  final String? dailyTaskInstanceId;
  final String? goalId;
  final DateTime createdAt;

  const IncomeEntry({
    required this.id,
    required this.amount,
    required this.currency,
    required this.source,
    required this.entryDate,
    required this.note,
    required this.dailyTaskInstanceId,
    required this.goalId,
    required this.createdAt,
  });

  factory IncomeEntry.fromJson(Map<String, dynamic> json) => IncomeEntry(
        id: json['id'] as String,
        amount: double.parse(json['amount'].toString()),
        currency: json['currency'] as String,
        source: json['source'] as String,
        entryDate: DateTime.parse(json['entry_date'] as String),
        note: json['note'] as String?,
        dailyTaskInstanceId: json['daily_task_instance_id'] as String?,
        goalId: json['goal_id'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class FinanceSummary {
  final double totalAllTime;
  final double totalThisMonth;
  final double unallocatedAmount;

  const FinanceSummary({
    required this.totalAllTime,
    required this.totalThisMonth,
    required this.unallocatedAmount,
  });

  factory FinanceSummary.fromJson(Map<String, dynamic> json) => FinanceSummary(
        totalAllTime: double.parse(json['total_all_time'].toString()),
        totalThisMonth: double.parse(json['total_this_month'].toString()),
        unallocatedAmount: double.parse(json['unallocated_amount'].toString()),
      );
}
